PR: https://github.com/odoo/odoo/pull/233963

From: c20d2cda2e292aadb673824c8179359e34c0daa1
From: Aurélien Warnon
Date: 2026-02-14 01:01:28

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = ['utm.mixin']

Total Changes: 194

[IMP] mass_mailing[_sms]: improve and generalize mailing UTM trackers

RATIONALE

The way we currently use the UTM trackers (aka source/medium/campaign) does not
comply with the industry standards, leading to confusion and globally unclear
or not usable reporting on UTM data.

PURPOSE

Use the recently introduced utm_reference logic to replace the current utm
source when creating link.trackers for mass mailings (and mass sms).

Also specify and generalize the UTM medium and source to complement reporting.

SPECS

We no longer create an utm.source every time we create a mailing and instead
directly set the mailing record as our utm reference.

See commit introducing the utm reference concept for more details.

EXAMPLE

Example of outcome, let's say on a sale.order created after clicking on a
button from a Black Friday campaign:
* UTM Campaign: Black Friday Sales
* UTM Medium: utm.utm_medium_email
* UTM Source: utm.utm_source_mailing
* UTM Reference: mailing.mailing,42

NAME REMOVAL

A side effect of removing the inheritance to utm.source.mixin is that it
removes the "name" field. It was globally unused and use cases have been
adapted to use "subject" instead.

The only exception is the A/B Testing use case where people may want to
identify different versions of the mailing records that will share the same
subject.
For that use case, a new field "A/B Testing Version" was introduced.

TECHNICAL

4 queries are added during mass mailing send.
These are due to the fact that the UTM source is now computed instead of being
created along with the mass.mailing record (old utm.mixin behavior).

Task-3330036

Part-of: odoo/odoo#233963
Related: odoo/enterprise#98569
Related: odoo/upgrade#8746
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/mass_mailing_sms/models/__init__.py
+++ b/addons/mass_mailing_sms/models/__init__.py
@@ -8,4 +8,5 @@ from . import mailing_trace
 from . import res_users
 from . import sms_sms
 from . import sms_tracker
-from . import utm
+from . import utm_campaign
+from . import utm_mixin

--- a/addons/mass_mailing_sms/models/mailing_mailing.py
+++ b/addons/mass_mailing_sms/models/mailing_mailing.py
@@ -59,12 +59,20 @@ class MailingMailing(models.Model):
 
     @api.depends('mailing_type')
     def _compute_medium_id(self):
-        super()._compute_medium_id()
         for mailing in self:
-            if mailing.mailing_type == 'sms' and (not mailing.medium_id or mailing.medium_id == self.env['utm.mixin']._utm_ref('utm.utm_medium_email')):
-                mailing.medium_id = self.env['utm.mixin']._utm_ref('utm.utm_medium_sms').id
-            elif mailing.mailing_type == 'mail' and (not mailing.medium_id or mailing.medium_id == self.env['utm.mixin']._utm_ref('utm.utm_medium_sms')):
-                mailing.medium_id = self.env['utm.mixin']._utm_ref('utm.utm_medium_email').id
+            if mailing.mailing_type == 'mail':
+                if not mailing.medium_id or mailing.medium_id == self.env['utm.mixin']._utm_ref('mass_mailing_sms.utm_medium_sms'):
+                    mailing.medium_id = self.env['utm.mixin']._utm_ref('utm.utm_medium_email').id
+            if mailing.mailing_type == 'sms':
+                if not mailing.medium_id or mailing.medium_id == self.env['utm.mixin']._utm_ref('utm.utm_medium_email'):
+                    mailing.medium_id = self.env['utm.mixin']._utm_ref('mass_mailing_sms.utm_medium_sms').id
+
+    @api.depends('mailing_type')
+    def _compute_source_id(self):
+        super()._compute_source_id()
+        for mailing in self:
+            if mailing.mailing_type == 'sms' and not mailing.source_id:
+                mailing.source_id = self.env['utm.mixin']._utm_ref('mass_mailing_sms.utm_source_mass_sms').id
 
     @api.depends('sms_template_id', 'mailing_type')
     def _compute_body_plaintext(self):

--- a/addons/mass_mailing_sms/models/utm.py
+++ b/addons/mass_mailing_sms/models/utm_campaign.py
@@ -1,9 +1,7 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from collections import defaultdict
-from odoo import _, api, fields, models
-from odoo.exceptions import UserError
+from odoo import api, fields, models
 
 
 class UtmCampaign(models.Model):
@@ -24,7 +22,6 @@ class UtmCampaign(models.Model):
         ('manual', 'Manual'),
         ('clicks_ratio', 'Highest Click Rate')], string="SMS Winner Selection", default="clicks_ratio")
 
-
     @api.depends('mailing_sms_ids')
     def _compute_mailing_sms_count(self):
         mailing_sms_data = self.env['mailing.mailing']._read_group(
@@ -75,11 +72,3 @@ class UtmCampaign(models.Model):
                 continue
             ab_testing_mailings.action_send_winner_mailing()
         return ab_testing_campaign
-
-
-class UtmMixin(models.Model):
-    _inherit = ['utm.mixin']
-
-    @property
-    def SELF_REQUIRED_UTM_REF(self):
-        return super().SELF_REQUIRED_UTM_REF | {'mass_mailing_sms.utm_medium_sms': ('SMS', 'utm.medium')}

--- a/None
+++ b/addons/mass_mailing_sms/models/utm_mixin.py
@@ -0,0 +1,14 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+
+
+class UtmMixin(models.AbstractModel):
+    _inherit = 'utm.mixin'
+
+    @property
+    def SELF_REQUIRED_UTM_REF(self):
+        return super().SELF_REQUIRED_UTM_REF | {
+            'mass_mailing_sms.utm_medium_sms': ('SMS', 'utm.medium'),
+            'mass_mailing_sms.utm_source_mass_sms': ('Mass SMS', 'utm.source'),
+        }
